### PR TITLE
Add support for port mapping in docker hints

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
Add support for port mapping in docker autodiscover. The idea is similar to what we do for Kubernetes which was added at https://github.com/elastic/beats/pull/19398. 

The idea is to make it possible for collecting metrics using the HostPort of the container as described at https://github.com/elastic/beats/issues/19825, by referring to the proper internal port of the container (see manual testing notes for an example).

## Why is it important?
To cover cases like https://github.com/elastic/beats/issues/19825.

## How to test this PR locally

1. Use the following configuration:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      hints.enabled: true

output.console:
  pretty: true
```
2. Run metricbet with `./metricbeat -e -d "module" -c metricbeat.yml`
3. Start a target container with `docker run --name nats --rm -p 4222:4222 -p 8081:8222 --label co.elastic.metrics/module="nats" --label co.elastic.metrics/metricsets="stats" --label co.elastic.metrics/hosts='localhost:${data.ports.8222}' nats:2.1.4 --http_port 8222`
4. Verify in the console output that Metricbeat is collecting metrics properly from `localhost:8222` endpoint.

## Related issues

- Closes https://github.com/elastic/beats/issues/19825
- Relates https://github.com/elastic/beats/pull/19398

### CODE REVIEW

**Review by jsoriano — COMMENTED** [review]
Nice. LGTM.

Please add a changelog entry.

And would it be possible to add tests for this?

**Comment by ChrsMark** [comment]
> Nice. LGTM.
> 
> Please add a changelog entry.
> 
> And would it be possible to add tests for this?

Yeap, I'm planning to add docs/changelog and good idea adding tests too!

File: libbeat/tests/docker/docker.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func NewClient() (Client, error) {
 }
 
 // ContainerStart pulls and starts the given container
-func (c Client) ContainerStart(image string, cmd []string, labels map[string]string) (string, error) {
+func (c Client) ContainerStart(image string, cmd []string, labels map[string]string, hostConfig *container.HostConfig) (string, error) {
```
Comment (jsoriano) [inline libbeat/tests/docker/docker.go:46]: Nit. Consider defining new method to keep backwards compatibility. Though probably nothing is using this method outside of beats.
```suggestion
// ...
type StartOptions struct {
    HostConfig *container.HostConfig
}

// ...
func ContainerStart(image string, cmd []string, labels map[string]string) (string, error) {
    return ContainerStartWithOptions(image, cmd, labels, StartOptions{})
}

// ...
func (c Client) ContainerStartWithOptions(image string, cmd []string, labels map[string]string, options StartOptions) (string, error) {
...
```

**Review by jsoriano — APPROVED** [review]

File: libbeat/tests/docker/docker.go (hunk 1 above)
Comment (ChrsMark) [reply libbeat/tests/docker/docker.go]: Thank you! This test is not correct nevertheless so I will revert this. Unfortunately we don't have proper tests at the moment that I could extend to cover this feature addition.

**Review by MichaelKatsoulis — APPROVED** [review]
LGTM

**Pull request merged**